libyul/YulString.h

```diff
@@ -117,10 +130,15 @@ class YulString
 	bool operator!=(YulString const& _other) const { return m_handle.id != _other.m_handle.id; }
 
 	bool empty() const { return m_handle.id == 0; }
-	std::string const& str() const
+	std::string const str() const
```

Comment:
Actually, although @chriseth wrote in the issue that the suffix should *not* be turned into a string, I'm not sure whether this here as a result is all that good. Wasn't one of the main reasons for the ``YulStringRepository`` to *avoid* string allocations? Returning a newly constructed string by value will allocate a new string every time ``str`` is called... I think the suffix *should* in fact be added to the string and in addition to the suffix the length of the prefix should be stored in the repository. That way we can reuse the string in the repository here (and return by reference again) and still return (a copy of) the prefix below (in the future once we support C++17 we can even use a ``stringview`` for the prefix, avoiding the copy).

@chriseth Do you agree with that?

Reply:
The question is what is done more often: Creating a new suffixed string or using `str()`. Would it be possible to only convert the number to a string at the first time that is needed? Perhaps we could also profile both solutions.

Reply:
Maybe we should turn the ``std::pair<std::string, size_t>`` that ``m_strings`` maps to now to a proper ``struct`` or member ``class`` of the repository - then we could easily create some more caching logic in it.
It could e.g. always store the prefix as string and suffix as number and additionally have an ``optional`` (resp. ``unique_ptr`` or ``shared_ptr``) to store the full string on demand. That would waste a bit of memory, but keep allocations and copies (close) to a minimum.

Reply:
That way both ``idToString`` and ``idToPrefix`` could be kept in the repository (and return const references) and ``str()`` and ``prefix()`` could again just forward to them (and return const references as well).
``idToString`` in the repository would forward to a member of the new ``struct`` which would allocate the full string on demand.

@Altrum Do you want to implement that? (I'm not sure if my description is clear enough - if not I could push add a commit with those changes to the PR myself)

Reply:
We should really profile that. The reason why we got the speedup from string to YulString might very well not be due to the fact that we save the comparison loop, but rather that the YulString is fully kept on the stack and does not have to be loaded from memory.

Reply:
Yes, we should definitely profile all of that - but I'd be surprised, if preventing the allocation of new strings in ``str()`` while not increasing the size of ``YulString`` were detrimental in the end.

Reply:
I think I understand what you mean--that it should basically only become a "full string" the first time `str()` is called. I'm not sure how to do profiling for this tho 😅.

Also, would I just do another commit if I wanted to implement this? I've been amending my commit instead of adding new ones.